includes/class-wc-structured-data.php

```diff
@@ -0,0 +1,202 @@
+    
+    if ( $product_count === 1 ) {
+      $structured_data = isset( $reviews ) ? $products[0] + array( 'review' => $reviews ) : $products[0];
+    } elseif ( $product_count > 1 ) {
+      $structured_data = array( '@graph' => $products );
+    }
+    
+    if ( ! isset( $structured_data ) ) {
+      return array();
+    }
+    
+    $context['@context'] = 'http://schema.org/';
+
+    return $context + $structured_data;
+  }
+
+  /**
+   * Contructor
+   */
+  public function __construct() {
+    add_action( 'woocommerce_before_shop_loop_item', array( $this, 'generate_product_category_data' ) );
+    add_action( 'woocommerce_single_product_summary', array( $this, 'generate_product_data' ) );
+    add_action( 'woocommerce_review_meta', array( $this, 'generate_product_review_data' ) );
+    add_action( 'wp_footer', array( $this, 'enqueue_data' ) );
+  }
+
+  /**
+   * Sanitizes, encodes and echoes the structured data into `wp_footer` action hook.
+   */
+  public function enqueue_data() {
+    if ( $structured_data = $this->get_data() ) {
+
+      array_walk_recursive( $structured_data, array( $this, 'sanitize_data' ) );
+
+      echo '<script type="application/ld+json">' . wp_json_encode( $structured_data ) . '</script>';
+    }
+  }
+
+  /**
+   * Callback function for sanitizing the structured data.
+   * Can't use anonymous function due to PHP v5.2 incompatibility...
+   *
+   * @param ref
+   */
+  private function sanitize_data( &$value ) {
+    $value = sanitize_text_field( $value );
+  }
+
+  /**
+   * Generates the product category structured data...
+   * Hooked into the `woocommerce_before_shop_loop_item` action hook...
+   */
+  public function generate_product_category_data() {
+    if ( ! is_product_category() ) {
+      return;
+    }
+    
+    $this->generate_product_data();
+  }
+  
+  /**
+   * Generates the product structured data...
+   * Hooked into the `woocommerce_single_product_summary` action hook...
+   * Applies the `woocommerce_product_structured_data` filter hook for clean structured data customization...
+   */
+  public function generate_product_data() {
+    global $product;
+
+    if ( $product->is_downloadable() ) {
+      switch ( $product->download_type ) {
+        case 'application' :
+          $type = "SoftwareApplication";
+        break;
+        case 'music' :
+          $type = "MusicAlbum";
+        break;
+        default :
+          $type = "Product";
+        break;
+      }
+    } else {
+      $type = "Product";
+    }
+
+    $json['@type']             = $type;
+    $json['@id']               = get_the_permalink();
+    $json['name']              = get_the_title();
+    $json['image']             = wp_get_attachment_url( $product->get_image_id() );
+    $json['description']       = get_the_excerpt();
+    $json['url']               = get_the_permalink();
+    $json['sku']               = $product->get_sku();
+
+    if ( $brand = $product->get_attribute( __( 'brand', 'woocommerce' ) ) ) {
```

Comment:
Better remove, people can add by the filter, since is part of WooCommerce core.
This can comes from a extension like: https://woocommerce.com/products/brands/